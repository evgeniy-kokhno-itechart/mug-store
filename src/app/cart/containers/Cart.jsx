import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import CartTableConnected from './CartTable';
import TotalCost from '../components/TotalCost';
import { cartCostsSelector } from '../CartSelectors';

export class Cart extends Component {
  renderCartDetails(products) {
    const { currentUserName, cart } = this.props;
    return (
      <React.Fragment>
        <CartTableConnected
          sortColumn="id"
          productsInCart={products}
          onDeleteFromCart={this.handleDeleteFromCart}
          onIncrementClick={this.handleIncrementClick}
          onDecrementClick={this.handleDecrementClick}
          onQuantityChange={this.handleQuantityChange}
        />
        <div className="row justify-content-between mx-2">
          <TotalCost total={cart.totalCost} customClasses="h5" />
          <div>
            <Link className="btn btn-secondary justify-content-end" to={currentUserName ? '/order/mycart' : '/order'}>
              Order Now!
            </Link>
          </div>
        </div>
      </React.Fragment>
    );
  }

  renderEmptyMessage() {
    return <h1 className="m-3 text-center">Your cart is empty</h1>;
  }

  render() {
    const { cart } = this.props;

    return cart.products.length ? this.renderCartDetails(cart.products) : this.renderEmptyMessage();
  }
}

Cart.propTypes = {
  cart: PropTypes.shape({
    products: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        imageURL: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        currentCurrencyPrice: PropTypes.number,
        quantity: PropTypes.number,
      }),
    ),
    totalCost: PropTypes.number,
  }).isRequired,

  currentUserName: PropTypes.string,
};

Cart.defaultProps = {
  currentUserName: '',
};

const mapStateToProps = state => ({
  cart: cartCostsSelector(state),
  currentUserName: state.user.currentUser.name,
});

export default connect(mapStateToProps)(Cart);
