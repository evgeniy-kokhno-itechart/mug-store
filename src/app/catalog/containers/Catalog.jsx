/* eslint-disable no-nested-ternary */
import React, { Component } from 'react';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { categoryActions } from '../CategoryActions';
import paginate from '../Paginate';
import CatalogTable from '../components/CatalogTable';
import CatalogTableHeader from '../components/CatalogTableHeader';
import CatalogTableFooter from '../components/CatalogTableFooter';
import { ListGroup, ErrorMessage, Spinner } from '../../shared';
import { // deleteProduct, getProducts,
  productsPricesSelector,
  productsActions,
} from '../../product';
import { cartActions } from '../../cart';

export class Catalog extends Component {
  state = {
    searchQuery: '',
    sortColumn: { id: 'title_asc', path: 'title', order: 'asc' },
    pageSize: 10,
    currentPage: 1,
  };

  pageSizeOptions = [{ id: '5', name: '5' }, { id: '10', name: '10' }, { id: '15', name: '15' }];

  sortOptions = [
    { id: 'currentCurrencyPrice_asc', name: 'Price A-Z' },
    { id: 'currentCurrencyPrice_desc', name: 'Price Z-A' },
    { id: 'title_asc', name: 'Title A-Z' },
    { id: 'title_desc', name: 'Title Z-A' },
    { id: 'rate_asc', name: 'Rate 1-5' },
    { id: 'rate_desc', name: 'Rate 5-1' },
  ];

  allProductsCategory = { id: '', name: 'All Products' };

  componentDidMount() {
    this.props.getProducts();
  }

  displayIsLoading = () => {
    const { products, category } = this.props;
    if (products.catalogProductsStatus.isInProcess || category.loadingStatus.isInProcess || products.productDeletingStatus.isInProcess) {
      return <Spinner spinnerClasses="spinner--big" wrapperClasses="mt-5" />;
    }
    return null;
  }

  displayGotError = () => {
    const { products, category } = this.props;
    if (products.catalogProductsStatus.hasFailed || category.loadingStatus.hasFailed || products.productDeletingStatus.hasFailed) {
      return (
        <ErrorMessage message={products.catalogProductsStatus.error || category.loadingStatus.error || category.loadingStatus.error} />
      );
    }
    return null;
  }

  displayPagedDataOrEmpty = () => {
    const { totalCount, productsOnPage } = this.getPagedData();
    return productsOnPage.length
      ? this.renderProducts(totalCount, productsOnPage)
      : this.renderEmptyMessage();
  }

  getPagedData = () => {
    const {
      pageSize, currentPage, sortColumn, searchQuery,
    } = this.state;

    const { category } = this.props;

    const allProducts = this.props.products.productsForCatalog;
    let filtered = allProducts;
    if (searchQuery) {
      filtered = allProducts.filter(m => m.title.toLowerCase().includes(searchQuery.toLowerCase()));
    } else if (category.currentCategory && category.currentCategory.id) {
      filtered = allProducts.filter(m => m.category.id === category.currentCategory.id);
    }

    const sorted = _.orderBy(filtered, [sortColumn.path], [sortColumn.order]);

    const productsOnPage = paginate(sorted, currentPage, pageSize);

    return { totalCount: filtered.length, productsOnPage };
  };

  handleItemSelect = (category) => {
    this.setState({ searchQuery: '', currentPage: 1 });
    this.props.changeCategory(category);
  };

  handlePageChange = (page) => {
    this.setState({ currentPage: page });
  };

  handleSearch = (e, query) => {
    e.preventDefault();
    this.props.changeCategory(this.allProductsCategory);
    this.setState({ searchQuery: query, currentPage: 1 });
  };

  handleSort = (e) => {
    const detailsToBeSplitted = e.currentTarget.value;
    if (detailsToBeSplitted) {
      const sortInfo = detailsToBeSplitted.split('_');

      this.setState({ sortColumn: { id: detailsToBeSplitted, path: sortInfo[0], order: sortInfo[1] } });
    } else {
      this.setState({ sortColumn: { path: 'title', order: 'asc' } });
    }
  };

  handleItemsCountChange = (e) => {
    const pageSizeString = e.currentTarget.value;
    const pageSize = +pageSizeString;
    this.setState({ pageSize, currentPage: 1 });
  };

  handleDelete = (productId) => {
    this.props.deleteProduct(productId);
    this.props.getProducts();
  };

  renderEmptyMessage = () => (
    <h1 className="m-2 text-center">
      There are no products in the catalog
      <div className="m-1 m-md-3 mx-auto">
        <Link className="btn btn-secondary" to="/">
          To Main Page
        </Link>
      </div>
    </h1>
  );

  renderProducts = (totalCount, productsOnPage) => {
    const {
      sortColumn, searchQuery, pageSize, currentPage,
    } = this.state;

    const {
      // eslint-disable-next-line no-shadow
      currentUserRoles, category, isCurrencyLoading, addToCart,
    } = this.props;

    return (
      <div className="row mt-3">
        <div className="col-11 col-sm-3 col-lg-2 mx-auto mb-2">
          <ListGroup
            items={[this.allProductsCategory, ...category.categories]}
            selectedItem={category.currentCategory}
            onItemSelect={this.handleItemSelect}
          />
        </div>

        <div className="col-12 col-sm-9 col-lg-10">
          <CatalogTableHeader
            currentUserRoles={currentUserRoles}
            searchQuery={searchQuery}
            sortColumnKey={sortColumn.id}
            sortOptions={this.sortOptions}
            handleSearch={this.handleSearch}
            handleSort={this.handleSort}
          />

          <CatalogTable
            sortColumn={sortColumn}
            productsOnPage={productsOnPage}
            currentUserRoles={currentUserRoles}
            isCurrencyLoading={isCurrencyLoading}
            addToCart={addToCart}
            onDelete={this.handleDelete}
          />

          <CatalogTableFooter
            totalCount={totalCount}
            pageSize={pageSize}
            pageSizeOptions={this.pageSizeOptions}
            currentPage={currentPage}
            onPageChange={this.handlePageChange}
            onItemsCountChange={this.handleItemsCountChange}
          />
        </div>
      </div>
    );
  };

  render() {
    return this.displayIsLoading() || this.displayGotError() || this.displayPagedDataOrEmpty();
  }
}

Catalog.propTypes = {
  currentUserRoles: PropTypes.arrayOf(PropTypes.string),

  products: PropTypes.shape({
    productsForCatalog: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        imageURL: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        category: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
        basePrice: PropTypes.number,
        currentCurrencyPrice: PropTypes.number,
        discount: PropTypes.number,
        producer: PropTypes.string,
        publishDate: PropTypes.string,
        rate: PropTypes.string,
      }),
    ),
    catalogProductsStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
    productDeletingStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
  }),

  category: PropTypes.shape({
    categories: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })),
    currentCategory: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
    loadingStatus: PropTypes.shape({ isInProcess: PropTypes.bool, hasFailed: PropTypes.bool, error: PropTypes.string }),
  }),

  isCurrencyLoading: PropTypes.bool,

  getProducts: PropTypes.func.isRequired,
  deleteProduct: PropTypes.func.isRequired,
  changeCategory: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
};

Catalog.defaultProps = {
  currentUserRoles: [],

  products: {
    productsForCatalog: [],
    productsGettingStatus: { isInProcess: false, hasFailed: false, error: '' },
    productDeletingStatus: { isInProcess: false, hasFailed: false, error: '' },
  },

  category: {
    categories: [],
    currentCategory: {},
    loadingStatus: { isInProcess: false, hasFailed: false, error: '' },
  },

  isCurrencyLoading: true,
};

const mapStateToProps = state => ({
  currentUserRoles: state.user.currentUser.roles,

  products: {
    productsForCatalog: productsPricesSelector(state),
    catalogProductsStatus: state.products.catalogProductsStatus,
    productDeletingStatus: state.products.deletingStatus,
  },

  isCurrencyLoading: state.currency.currenciesStatus.isInProcess,

  category: state.category,
});

const mapDispatchToProps = {
  changeCategory: categoryActions.ChangeCategory,
  getProducts: productsActions.GetProducts.InitiateApiCall,
  deleteProduct: productsActions.DeleteProduct.InitiateApiCall,
  addToCart: cartActions.AddToCart,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Catalog);
