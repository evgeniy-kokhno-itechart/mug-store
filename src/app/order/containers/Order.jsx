import React from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import OrderTable from '../components/OrderTable';
import OrderForm from './OrderForm';
import { TotalCost, submitCartOrder, cartCostsSelector } from '../../cart';

export const Order = ({
  cart,
  currentUser,
  isCurrencyLoading,
  // eslint-disable-next-line no-shadow
  submitCartOrder,
}) => (
  <React.Fragment>
    <h1 className="text-center m-3">Please check your order</h1>
    <div className="row m-2">
      <div className="col-md-6">
        <OrderTable sortColumn="title" products={cart.products} isCurrencyLoading={isCurrencyLoading} />
        <TotalCost total={cart.totalCost} customClasses="float-right" />
      </div>
      <div className="col-md-5 offset-md-1">
        <OrderForm currentUser={currentUser} onOrderSubmit={submitCartOrder} />
      </div>
    </div>
  </React.Fragment>
);

Order.propTypes = {
  cart: PropTypes.shape({
    products: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        quantity: PropTypes.number,
        price: PropTypes.object,
        title: PropTypes.string,
      }),
    ),
    totalCost: PropTypes.number,
  }).isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.string,
    address: PropTypes.string,
    city: PropTypes.string,
    country: PropTypes.string,
    name: PropTypes.string,
    phone: PropTypes.string,
    roles: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string,
  }),

  isCurrencyLoading: PropTypes.bool,

  submitCartOrder: PropTypes.func.isRequired,
};

Order.defaultProps = {
  currentUser: {},
  isCurrencyLoading: true,
};

const mapStateToProps = state => ({
  cart: cartCostsSelector(state),
  currentUser: state.user.currentUser,
  isCurrencyLoading: state.currency.currenciesStatus.isInProcess,
});

const mapDipatchToProps = {
  submitCartOrder,
};

export default connect(
  mapStateToProps,
  mapDipatchToProps,
)(Order);
