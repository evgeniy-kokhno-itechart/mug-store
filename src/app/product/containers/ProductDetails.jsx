/* eslint-disable no-nested-ternary */
import React, { Component } from 'react';
import ImageGallery from 'react-image-gallery';
import _ from 'lodash';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import getProductImageURLs from '../ImageService';
import { productCostSelector } from '../ProductsSelectors';
import { getProduct, clearCurrentProductInfo } from '../ProductsActions';
import { InformationItem, ErrorMessage, Spinner } from '../../shared';
import { addToCart } from '../../cart';
import { BuyNowButton, ToCatalogButton } from '../../catalog';

export class ProductDetails extends Component {
  state = {
    imageURLs: [],
  };

  fields = [
    { label: 'Title', path: 'title' },
    { label: 'Description', path: 'description' },
    { label: 'Category', path: 'category.name' },
    { label: 'Producer', path: 'producer' },
    { label: 'Price', path: 'currentCurrencyPrice' },
  ];

  componentDidMount() {
    const { match } = this.props;

    const productId = match.params.id;

    this.props.getProduct(productId);

    //  imageURLs were separated from original model for sample purposes since they are the same across all products
    //  my-json-server.typicode.com char limit 10000 has been reached thus URLs data couldn't be placed there
    const imageURLs = getProductImageURLs(productId);
    this.setState({ imageURLs });
  }

  getInfo = (item, path) => _.get(item, path);

  conponentWillUnmount() {
    this.props.clearCurrentProductInfo();
  }

  renderFieldInfo = (field) => {
    const { product } = this.props;

    let fieldInfo = null;
    if (!field.content) {
      fieldInfo = this.getInfo(product, field.path);
    } else {
      fieldInfo = field.content(product);
    }
    return <InformationItem key={field.label} label={field.label} info={fieldInfo} />;
  };

  renderSpinner = () => <Spinner customSizeClassName="product-details__spinner" marginBootstrapClassName="mt-5" />;

  renderErrorMessage = message => <ErrorMessage message={message} />;

  renderProductDetails = () => {
    const { imageURLs } = this.state;
    const { product } = this.props;

    return (
      <React.Fragment>
        <h1 className="display-5 text-center m-3">{`${product.title} Details`}</h1>
        {imageURLs && (
          <div className="col-10 col-md-9 col-lg-7 mx-auto">
            <ImageGallery
              items={imageURLs}
              infinite={false}
              showPlayButton={false}
              showBullets={false}
              showFullscreenButton={false}
              useBrowserFullscreen={false}
            />
          </div>
        )}

        {this.fields.map(field => this.renderFieldInfo(field))}

        <div className="d-flex m-3">
          <ToCatalogButton />
          <BuyNowButton customClasses="ml-auto" onBuyNow={this.props.addToCart} product={product} />
        </div>
      </React.Fragment>
    );
  };

  render() {
    return this.props.isProductLoading
      ? this.renderSpinner()
      : this.props.hasProductLoadingFailed
        ? this.renderErrorMessage(this.props.errorWhileLoading)
        : this.renderProductDetails();
  }
}

ProductDetails.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    category: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
    description: PropTypes.string,
    producer: PropTypes.string,
    currentCurrencyPrice: PropTypes.number,
  }),
  isProductLoading: PropTypes.bool.isRequired,
  hasProductLoadingFailed: PropTypes.bool,
  errorWhileLoading: PropTypes.string,
  match: PropTypes.shape({ params: PropTypes.shape({ id: PropTypes.string }) }).isRequired,

  addToCart: PropTypes.func.isRequired,
  getProduct: PropTypes.func.isRequired,
  clearCurrentProductInfo: PropTypes.func.isRequired,
};

ProductDetails.defaultProps = {
  product: {
    id: '',
    imageURL: '',
    title: 'defaultProduct',
    description: '',
    category: {},
    basePrice: 0,
    currentCurrencyPrice: 0,
    discount: 0,
    producer: '',
    rate: '',
  },
  hasProductLoadingFailed: false,
  errorWhileLoading: '',
};

const mapStateToProps = state => ({
  product: productCostSelector(state),
  isProductLoading: state.products.currentProductStatus.isGettingByIdInProcess,
  hasProductLoadingFailed: state.products.currentProductStatus.hasGettingByIdFailed,
  errorWhileLoading: state.products.currentProductStatus.error,
});

const mapDispatchToProps = {
  addToCart,
  getProduct,
  clearCurrentProductInfo,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProductDetails);
