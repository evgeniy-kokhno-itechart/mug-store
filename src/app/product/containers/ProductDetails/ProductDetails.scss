.prod-details{
    &__spinner {
        margin-top: 2rem;
    }

    &__header {
        text-align: center;
        margin: 2rem;
    }

    &__images {
        width: 50%;
        margin: 0 auto;
    }

    &__footer {
        display: flex;
        justify-content: space-between;
        margin: 1rem;
    }
}