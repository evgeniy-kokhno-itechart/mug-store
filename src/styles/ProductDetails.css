#bootstrap-overrides .product-details__spinner {
  width: 3rem;
  height: 3rem;
}
