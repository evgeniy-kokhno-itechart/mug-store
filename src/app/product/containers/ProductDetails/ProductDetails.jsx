/* eslint-disable no-nested-ternary */
import React, { Component } from 'react';
import ImageGallery from 'react-image-gallery';
import _ from 'lodash';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import getProductImageURLs from '../../ImageService';
import { productCostSelector } from '../../ProductsSelectors';
import { productsActions } from '../../ProductsActions';
import { InformationItem, ErrorMessage, Spinner } from '../../../shared';
import { cartActions } from '../../../cart';
import { BuyNowButton, ToCatalogButton } from '../../../catalog';
import './ProductDetails.scss';

export class ProductDetails extends Component {
  state = {
    imageURLs: [],
  };

  fields = [
    { label: 'Title', path: 'title' },
    { label: 'Description', path: 'description' },
    { label: 'Category', path: 'category.name' },
    { label: 'Producer', path: 'producer' },
    { label: 'Price', path: 'currentCurrencyPrice' },
  ];

  componentDidMount() {
    const { match } = this.props;

    const productId = match.params.id;

    this.props.getProduct(productId);

    //  imageURLs were separated from original model for sample purposes since they are the same across all products
    //  my-json-server.typicode.com char limit 10000 has been reached thus URLs data couldn't be placed there
    const imageURLs = getProductImageURLs(productId);
    this.setState({ imageURLs });
  }

  getInfo = (item, path) => _.get(item, path);

  conponentWillUnmount() {
    this.props.clearCurrentProductInfo();
  }

  renderFieldInfo = (field) => {
    const { productState } = this.props;

    let fieldInfo = null;
    if (!field.content) {
      fieldInfo = this.getInfo(productState.product, field.path);
    } else {
      fieldInfo = field.content(productState.product);
    }
    return <InformationItem key={field.label} label={field.label} info={fieldInfo} />;
  };

  renderSpinner = () => <Spinner spinnerClasses="spinner--medium" wrapperClasses="prod-details__spinner" />;

  renderErrorMessage = message => <ErrorMessage message={message} />;

  renderProductDetails = () => {
    const { imageURLs } = this.state;
    const { productState } = this.props;

    return (
      <div className="prod-details">
        <h1 className="prod-details__header">{`${productState.product.title} Details`}</h1>
        {imageURLs && (
          <div className="prod-details__images">
            <ImageGallery
              items={imageURLs}
              infinite={false}
              showPlayButton={false}
              showBullets={false}
              showFullscreenButton={false}
              useBrowserFullscreen={false}
            />
          </div>
        )}

        {this.fields.map(field => this.renderFieldInfo(field))}

        <div className="prod-details__footer">
          <ToCatalogButton />
          <BuyNowButton customClasses="" onBuyNow={this.props.addToCart} product={productState.product} />
        </div>
      </div>
    );
  };

  render() {
    const { productState } = this.props;
    return productState.productStatus.isInProcess
      ? this.renderSpinner()
      : productState.productStatus.hasFailed
        ? this.renderErrorMessage(productState.productStatus.error)
        : this.renderProductDetails();
  }
}

ProductDetails.propTypes = {
  productState: PropTypes.shape({
    product: PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      category: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
      description: PropTypes.string,
      producer: PropTypes.string,
      currentCurrencyPrice: PropTypes.number,
    }),
    productStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
  }),

  match: PropTypes.shape({ params: PropTypes.shape({ id: PropTypes.string }) }).isRequired,

  addToCart: PropTypes.func.isRequired,
  getProduct: PropTypes.func.isRequired,
  clearCurrentProductInfo: PropTypes.func.isRequired,
};

ProductDetails.defaultProps = {
  productState: {
    product: {
      id: '',
      imageURL: '',
      title: 'defaultProduct',
      description: '',
      category: {},
      basePrice: 0,
      currentCurrencyPrice: 0,
      discount: 0,
      producer: '',
      rate: '',
    },
    productStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
  },
};

const mapStateToProps = state => ({
  productState: {
    product: productCostSelector(state),
    productStatus: state.products.currentProductStatus,
  },
});

const mapDispatchToProps = {
  addToCart: cartActions.AddToCart,
  getProduct: productsActions.GetProduct.InitiateApiCall,
  clearCurrentProductInfo: productsActions.ClearCurrentProductInfo,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProductDetails);
