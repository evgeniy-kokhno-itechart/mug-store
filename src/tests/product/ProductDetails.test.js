/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { ProductDetails } from '../../product/containers/ProductDetails';

configure({ adapter: new Adapter() });

describe('<ProductDetails />', () => {
  let ProductDetailsWrapper;
  const mockProduct = {
    id: '1',
    imageURL: '#',
    title: 'Mug 1',
    description: 'product 1 description',
    category: { id: '1', name: 'category 1' },
    basePrice: 5,
    currentCurrencyPrice: 5,
    discount: 0,
    producer: 'test producer 1',
    publishDate: 'Tue Aug 20 2019 16:27:02 GMT+0300 (Moscow Standard Time)',
    rate: '5',
  };

  beforeEach(() => {
    const addToCart = jest.fn();
    const getProduct = jest.fn();
    const clearCurrentProductInfo = jest.fn();
    ProductDetailsWrapper = shallow(
      <ProductDetails
        isProductLoading
        match={{ params: { id: '100' } }}
        addToCart={addToCart}
        getProduct={getProduct}
        clearCurrentProductInfo={clearCurrentProductInfo}
      />,
    );
  });

  it('renders Spinner properly with default props', () => {
    expect(ProductDetailsWrapper).toMatchSnapshot();
  });

  it('renders product details once isProductLoading is succesfully done', () => {
    ProductDetailsWrapper.setProps({
      isProductLoading: false,
      hasProductLoadingFailed: false,
      errorWhileLoading: '',
      product: mockProduct,
    });
    expect(ProductDetailsWrapper).toMatchSnapshot();
  });

  it('renders ErrorMessage once hasProductLoadingFailed and errorWhileLoading provided', () => {
    ProductDetailsWrapper.setProps({ isProductLoading: false, hasProductLoadingFailed: true, errorWhileLoading: 'test error' });
    expect(ProductDetailsWrapper).toMatchSnapshot();
  });

  it('render field info correctly with renderFieldInfo method', () => {
    ProductDetailsWrapper.setProps({
      product: mockProduct,
      isProductLoading: false,
      hasProductLoadingFailed: false,
      errorWhileLoading: '',
    });
    ProductDetailsWrapper.instance().renderFieldInfo({ label: 'Category', path: 'category.name' });
    expect(ProductDetailsWrapper).toMatchSnapshot();
  });
});
