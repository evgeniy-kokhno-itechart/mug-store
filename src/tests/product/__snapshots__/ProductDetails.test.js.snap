// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> render field info correctly with renderFieldInfo method 1`] = `
<Fragment>
  <h1
    className="display-5 text-center m-3"
  >
    Mug 1 Details
  </h1>
  <div
    className="col-10 col-md-9 col-lg-7 mx-auto"
  >
    <ImageGallery
      autoPlay={false}
      disableArrowKeys={false}
      disableSwipe={false}
      disableThumbnailScroll={false}
      flickThreshold={0.4}
      indexSeparator=" / "
      infinite={false}
      isRTL={false}
      items={
        Array [
          Object {
            "original": "https://picsum.photos/700/300?image=1080",
            "thumbnail": "https://picsum.photos/250/150?image=1080",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=1068",
            "thumbnail": "https://picsum.photos/250/150?image=1068",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=998",
            "thumbnail": "https://picsum.photos/250/150?image=998",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=635",
            "thumbnail": "https://picsum.photos/250/150?image=635",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=431",
            "thumbnail": "https://picsum.photos/250/150?image=431",
          },
        ]
      }
      lazyLoad={false}
      preventDefaultTouchmoveEvent={false}
      renderFullscreenButton={[Function]}
      renderLeftNav={[Function]}
      renderPlayPauseButton={[Function]}
      renderRightNav={[Function]}
      showBullets={false}
      showFullscreenButton={false}
      showIndex={false}
      showNav={true}
      showPlayButton={false}
      showThumbnails={true}
      slideDuration={450}
      slideInterval={3000}
      startIndex={0}
      stopPropagation={false}
      swipeThreshold={30}
      swipingTransitionDuration={0}
      thumbnailPosition="bottom"
      useBrowserFullscreen={false}
      useTranslate3D={true}
    />
  </div>
  <InformationItem
    info="Mug 1"
    key="Title"
    label="Title"
  />
  <InformationItem
    info="product 1 description"
    key="Description"
    label="Description"
  />
  <InformationItem
    info="category 1"
    key="Category"
    label="Category"
  />
  <InformationItem
    info="test producer 1"
    key="Producer"
    label="Producer"
  />
  <InformationItem
    info={5}
    key="Price"
    label="Price"
  />
  <div
    className="d-flex m-3"
  >
    <ToCatalogButton
      customClasses=""
    />
    <BuyNowButton
      customClasses="ml-auto"
      onBuyNow={[MockFunction]}
      product={
        Object {
          "basePrice": 5,
          "category": Object {
            "id": "1",
            "name": "category 1",
          },
          "currentCurrencyPrice": 5,
          "description": "product 1 description",
          "discount": 0,
          "id": "1",
          "imageURL": "#",
          "producer": "test producer 1",
          "publishDate": "Tue Aug 20 2019 16:27:02 GMT+0300 (Moscow Standard Time)",
          "rate": "5",
          "title": "Mug 1",
        }
      }
    />
  </div>
</Fragment>
`;

exports[`<ProductDetails /> renders ErrorMessage once hasProductLoadingFailed and errorWhileLoading provided 1`] = `
<ErrorMessage
  message="test error"
/>
`;

exports[`<ProductDetails /> renders Spinner properly with default props 1`] = `
<Spinner
  customSizeClassName="product-details__spinner"
  marginBootstrapClassName="mt-5"
/>
`;

exports[`<ProductDetails /> renders product details once isProductLoading is succesfully done 1`] = `
<Fragment>
  <h1
    className="display-5 text-center m-3"
  >
    Mug 1 Details
  </h1>
  <div
    className="col-10 col-md-9 col-lg-7 mx-auto"
  >
    <ImageGallery
      autoPlay={false}
      disableArrowKeys={false}
      disableSwipe={false}
      disableThumbnailScroll={false}
      flickThreshold={0.4}
      indexSeparator=" / "
      infinite={false}
      isRTL={false}
      items={
        Array [
          Object {
            "original": "https://picsum.photos/700/300?image=1080",
            "thumbnail": "https://picsum.photos/250/150?image=1080",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=1068",
            "thumbnail": "https://picsum.photos/250/150?image=1068",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=998",
            "thumbnail": "https://picsum.photos/250/150?image=998",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=635",
            "thumbnail": "https://picsum.photos/250/150?image=635",
          },
          Object {
            "original": "https://picsum.photos/700/300?image=431",
            "thumbnail": "https://picsum.photos/250/150?image=431",
          },
        ]
      }
      lazyLoad={false}
      preventDefaultTouchmoveEvent={false}
      renderFullscreenButton={[Function]}
      renderLeftNav={[Function]}
      renderPlayPauseButton={[Function]}
      renderRightNav={[Function]}
      showBullets={false}
      showFullscreenButton={false}
      showIndex={false}
      showNav={true}
      showPlayButton={false}
      showThumbnails={true}
      slideDuration={450}
      slideInterval={3000}
      startIndex={0}
      stopPropagation={false}
      swipeThreshold={30}
      swipingTransitionDuration={0}
      thumbnailPosition="bottom"
      useBrowserFullscreen={false}
      useTranslate3D={true}
    />
  </div>
  <InformationItem
    info="Mug 1"
    key="Title"
    label="Title"
  />
  <InformationItem
    info="product 1 description"
    key="Description"
    label="Description"
  />
  <InformationItem
    info="category 1"
    key="Category"
    label="Category"
  />
  <InformationItem
    info="test producer 1"
    key="Producer"
    label="Producer"
  />
  <InformationItem
    info={5}
    key="Price"
    label="Price"
  />
  <div
    className="d-flex m-3"
  >
    <ToCatalogButton
      customClasses=""
    />
    <BuyNowButton
      customClasses="ml-auto"
      onBuyNow={[MockFunction]}
      product={
        Object {
          "basePrice": 5,
          "category": Object {
            "id": "1",
            "name": "category 1",
          },
          "currentCurrencyPrice": 5,
          "description": "product 1 description",
          "discount": 0,
          "id": "1",
          "imageURL": "#",
          "producer": "test producer 1",
          "publishDate": "Tue Aug 20 2019 16:27:02 GMT+0300 (Moscow Standard Time)",
          "rate": "5",
          "title": "Mug 1",
        }
      }
    />
  </div>
</Fragment>
`;
